tensorflow/contrib/periodic_intersperse/core/kernels/periodic_intersperse_op.h

```diff
@@ -0,0 +1,216 @@
+#ifndef TENSORFLOW_KERNELS_PERIODICINTERSPERSE_OP_H_
+#define TENSORFLOW_KERNELS_PERIODICINTERSPERSE_OP_H_
+
+#include <cmath>
+#include <type_traits>
+#include <vector>
+#include "tensorflow/core/lib/core/status.h"
+#include "tensorflow/core/framework/op.h"
+#include "tensorflow/core/framework/op_kernel.h"
+#include "tensorflow/core/framework/shape_inference.h"
+#include "tensorflow/core/framework/tensor_shape.h"
+
+using namespace tensorflow;
+using std::vector;
+using std::decay;
+
+
+template<class IndexVec_t, class Index_t>
+Index_t compute_input_index(IndexVec_t &Y, const Index_t &i,
+                            const IndexVec_t &X, const int &q,
+                            const vector<int64> &g, const vector<int64> &G,
+                            Index_t &result, vector<Index_t> &output_indices,
+                            const typename decay<decltype(Y.size())>::type &rank)
+{
+  result = 0;
+  output_indices.clear();
+
+  // un-rasterize the output index
+  {
+    auto last_reduced_i = i;
+    for (auto r = rank - 1;; --r) {
+      output_indices[r] = last_reduced_i % Y[r];
+      last_reduced_i = (last_reduced_i - output_indices[r]) / Y[r];
+      if (r == 0) break;
```

Comment:
are these for loops the same as `for (auto r = rank - 1; r > 0; --r)`?

If so could we put the end condition into the for loop to make it easier to read? Same below.